### PR TITLE
Added C API(GPU) to CMake and CircleCI

### PR DESCRIPTION
**Pull request by prabhat00155**

### CODE REVIEW

File: c_api/gpu/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+set(FAISS_C_SRC
+  DeviceUtils_c.cpp
+  GpuAutoTune_c.cpp
+  GpuClonerOptions_c.cpp
+  GpuIndex_c.cpp
+  GpuResources_c.cpp
+  StandardGpuResources_c.cpp
+)
+target_sources(faiss_c PRIVATE ${FAISS_C_SRC})
+target_link_libraries(faiss_c PRIVATE faiss)
```
Comment (beauby) [inline c_api/gpu/CMakeLists.txt:21]: The dependency from `faiss_c` to `faiss` should already be declared in the CPU part, so this is not necessary.

File: c_api/gpu/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+install(DIRECTORY ${PROJECT_SOURCE_DIR}
+        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/faiss/c_api/gpu/
+        FILES_MATCHING PATTERN "*.h")
+
+find_package(CUDAToolkit REQUIRED)
+find_package(OpenMP REQUIRED)
+target_link_libraries(faiss_c PRIVATE OpenMP::OpenMP_CXX)
```
[the remark this answers is not in the document]
Comment (prabhat00155) [reply c_api/gpu/CMakeLists.txt:28]: We don't add OpenMP dependency in `faiss_c's CMAKELIST`.

**Comment by Enet4** [comment]
I also get the same error as the one reported by CircleCI when trying this out locally.

```sh
cmake . -DFAISS_ENABLE_GPU=ON -DFAISS_ENABLE_C_API=ON
```

```none
-- The CXX compiler identification is GNU 10.2.0
-- The CUDA compiler identification is NVIDIA 11.2.67
-- The C compiler identification is GNU 10.2.0

CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_C_COMPILE_OBJECT
...

I'm afraid that I do not know how to fix this.

**Comment by prabhat00155** [comment]
> I also get the same error as the one reported by CircleCI when trying this out locally.
> 
> ```shell
> cmake . -DFAISS_ENABLE_GPU=ON -DFAISS_ENABLE_C_API=ON
> ```
> 
> ```
> -- The CXX compiler identification is GNU 10.2.0
> -- The CUDA compiler identification is NVIDIA 11.2.67
> -- The C compiler identification is GNU 10.2.0
> 
> ...
> 
> CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
> Missing variable is:
> CMAKE_C_COMPILE_OBJECT
> ```
> 
> I'm afraid that I do not know how to fix this.

This goes away if I add `project(faiss_c_gpu_library LANGUAGES C CXX)`.

File: c_api/gpu/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+cmake_minimum_required(VERSION 3.17 FATAL_ERROR)
+
+project(faiss_c_gpu_library LANGUAGES C CXX)
+
+set(CMAKE_C_STANDARD 11)
```
[the remark this answers is not in the document]
Comment (prabhat00155) [reply c_api/gpu/CMakeLists.txt]: Removing `project(faiss_c_gpu_library LANGUAGES C CXX)` gives this error:
```
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_C_COMPILE_OBJECT
```

**Comment by beauby** [comment]
@Enet4 I updated the wrappers for `index_cpu_to_gpu()` et al., would you mind taking a look?

**Review by Enet4 — COMMENTED** [review]
This looks good in general, including the latest changes. Thank you to everyone involved in the process of migrating the C API so far.

There is one possible tweak as I describe in a comment below, but I wouldn't require it.

My only major concern right now: how do we build a dynamic library? Existing consumers of this API are still expecting a dynamic link to `faiss_c`.

File: c_api/gpu/GpuResources_c.h [hunk 4]
```diff
@@ -69,6 +69,13 @@ int faiss_GpuResources_getAsyncCopyStreamCurrentDevice(
         FaissGpuResources*,
         cudaStream_t*);
 
+FAISS_DECLARE_CLASS(GpuResourcesProvider)
+
+FAISS_DECLARE_DESTRUCTOR(GpuResourcesProvider)
+
+int faiss_GpuResourcesProvider_getResources(FaissGpuResourcesProvider*,
```
[the remark this answers is not in the document]
Comment (beauby) [reply c_api/gpu/GpuResources_c.h:76]: Yes that was my main uncertainty about the changes, thanks for the pointer, will update.

**Comment by beauby** [comment]
@Enet4 The way to build shared libraries with CMake is to pass the `-DBUILD_SHARED_LIBS=ON` option to the `cmake` invocation. In the current setting, this will output `libfaiss.so` and `libfaiss_c.so`.

**Comment by Enet4** [comment]
That works, thanks!

No further comment, this PR looks good to me!

File: c_api/gpu/GpuResources_c.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (orihabAnyvision) [reply c_api/gpu/GpuResources_c.h:76]: Hi, I came across this thread by chance... I'm trying to write c bindings for some faiss gpu functionality (specifically Free gpu memory) and am running into the conversion problem @Enet4 hinted at I think. Im getting:

```
cannot use _cgo1 (type **_Ctype_struct_FaissGpuResourcesProvider_H) as type **_Ctype_struct_FaissGpuResources_H in argument to _Cfunc_faiss_GpuResourcesProvider_getResources
cannot use _cgo0 (type *_Ctype_struct_FaissGpuResourcesProvider_H) as type *_Ctype_struct_FaissGpuResources_H in argument to _Cfunc_faiss_GpuResources_free
```

I see in StandardGpuResource_c.h the classes are defined with FAISS_DECLARE_CLASS_INHERITED(StandardGpuResources, GpuResourcesProvider).. 
I see in GpuResources_c.cpp there is the following line of code:
```
reinterpret_cast<GpuResourcesProvider*>(res)->getResources()
```
which I dont really understand (I dont know cpp), but from the error I'm getting it seems like I have everything in order except for the **cast** part. 

Any insight as to how I could call the reinterpret_cast functionality from go?

**Pull request closed without merging**